### ISSUE TITLE
XAML Static Binding Source Fails With. x:DataType

### ISSUE THREAD

**Issue by david-maw** [dotnet/maui#26467]
### Description

When a page has an x:DataType set and an element on the page uses static binding the binding can produce incorrect results.

In my minimal example I have this XAML page definition

``` XAML
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:services="clr-namespace:DivisiBill.Services"
             xmlns:vm="clr-namespace:DivisiBill.ViewModels"
             x:DataType="vm:SettingsViewModel"
             Title="Program Settings"
             x:Class="DivisiBill.Views.SettingsPage">
     <Label IsVisible="{Binding Source={x:Static services:Utilities.FalseConstant}}"
            HorizontalOptions="Center" FontSize="Large" 
         Text="This text should not be visible"/>
</ContentPage>
```

And in Utilities.cs:

``` C#
namespace DivisiBill.Services;
public static class Utilities
{
    public const bool FalseConstant = false;
...
```
The result should be a mostly blank page, but it looks like this:
![Image](https://github.com/user-attachments/assets/d4ba056d-c433-4ea9-82c5-4384e8374bc4)

### Steps to Reproduce

1. Clone the repo
2. Build and debug it on windows
3. When the program comes up you should see the page shown in the description above.
4. If you remove the x:DataType specification at the page level it will then work correctly (the text will not show up).

### Link to public reproduction project repository

https://github.com/david-maw/Community-xaml-static.git

### Version with bug

9.0.12 SR1.2

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.100 SR10

### Affected platforms

Android, Windows, I was *not* able test on other platforms

### Affected platform versions

Windows 11, Android 14

### Did you find any workaround?

Add a property so you can use a Binding instead of a Static specification. 

### Relevant log output

```shell

```

**Comment by Bowman74** [comment]
A couple of thoughts:

- Is this flag turned on in the project? `<MauiEnableXamlCBindingWithSourceCompilation>true</MauiEnableXamlCBindingWithSourceCompilation>`

- Have you tried setting the type of binding for that specific line? i.e.
`<Label IsVisible="{Binding Source={x:Static services:Utilities.FalseConstant}, x:DataType=WhateverDataTypeservicesUtilitiesIs}"
            HorizontalOptions="Center" FontSize="Large" 
         Text="This text should not be visible"/>`

- Have you checked the Debug output window? Binding errors almost always show up there. I suspect this is happening because it is assuming that the type of your static binding for the label is `vm:SettingsViewModel` but it is not, it is something else but you didn't tell it what.

**Comment by david-maw** [comment]
Thanks @Bowman74, my basic complaint is that it used to work in .NET8 and it doesn't now. Worse, it fails silently, just giving the wrong answer. To your specific questions:

- Yes `MauiEnableXamlCBindingWithSourceCompilation` is set true in the project.
- Yes, I tried 'x:DataType="services:Utilities" and it didn't help but perhaps something else was called for.
- Yes, some sort of type mismatch seems likely to me too, but I saw nothing in the Output window, or the XAML Binding Errors window.

**Comment by StephaneDelcroix** [comment]
not discussing the bug here, but... I don't see the point of Binding to a x:Static. you could probably use x:Static directly.. or the value. that'd save cpu cycles and make your code easier to read/understand/debug

**Comment by Bowman74** [comment]
> another workaround would be to x:DataType=bool your Binding

@StephaneDelcroix would x:DataType=bool work when the base object is services:Utilities.FalseConstant? I.e. wouldn't the x:DataType need to correspond to whatever the base object is in the Source (in this case whatever Utilities is)? If not I'll have to do some more testing and alter the content of my course. IIRC when I tried that the binding wasn't compiled, or at least the warning still showed up that it wasn't compiled. But my memory may be faulty.

**Comment by david-maw** [comment]
@StephaneDelcroix usually yes, just using x:Static directly is better all round, but if you want some of the binding capabilities you're stuck with binding. For example (and I'll put this in a the sample):

``` XAML
        <Label HorizontalOptions="Center" FontSize="Medium" 
            Text="{Binding Source={x:Static services:Utilities.FalseValue},
            StringFormat='FalseValue = {0}',
            TargetNullValue='FalseValue evaluated to null'}"/>
``` 

Also, thanks @Bowman74 for reminding me about `MauiEnableXamlCBindingWithSourceCompilation `, that made my much simpler test case code fail as well, so I've been able to post a simpler test case.

**Comment by david-maw** [comment]
@StephaneDelcroix, I couldn't persuade x:DataType=bool to compile and in any case I agree with @Bowman74, wouldn't it actually be x:DataType=services:Utilities (that doesn't work either, but it does compile)?

**Comment by Ying-6** [comment]
This issue has been verified using Visual Studio 17.13.0 Preview 2.1 (9.0.12 $9.0.10). Can repro this issue on Windows and Android Platforms. And 8.0.100 works fine.

**Comment by StephaneDelcroix** [comment]
@david-maw I meant x:Boolean... sorry for the confusion

this works for me 

```
        IsVisible="{Binding Source={x:Static local:Maui26647.FalseConstant}, x:DataType=x:Boolean}"
```

**Comment by StephaneDelcroix** [comment]
@simonrozsival and I believe this works as expected, even if there's room for improvement at determining the DataType when the Source is set.

**Comment by david-maw** [comment]
Yep, @StephaneDelcroix , something similar works for me too, thanks for the correction. It seems like the trick is to provide the correct x:DataType. It would be nice if the XAML compiler were smart enough to do that itself, or at least warn the user, but nonetheless if you get the syntax right it clearly works.

**Comment by StephaneDelcroix** [comment]
progress on inferring the type from the Source is tracked here: #21834

**Comment by david-maw** [comment]
Thanks, in my case specifying x:DataType isn't especially arduous but when I get a minute I'll complain to the VS folk that it doesn't appear as an option in binding expressions.
